Refactoring: prepare for supporting different CashAddress prefixes

PiRK opened the pull request:
## Description
The current implementation of `CashAddress` has the `"bitcoincash:"` prefix hardcoded. This PR makes the code ready to support other CashAddress prefixes. This is a first necessary step for suppporting eCash.

This is achieved by refactoring the code in `cashaddresss.h/cpp`, to make `CashAddress` a base class capable of supporting different prefixes, and adding a subclass `BitcoinCashAddress` which is now used in the rest of the codebase where `CashAddress` was used before.

## Testing instructions

```
cmake -H. -Bbuild -DCMAKE_BUILD_TYPE=Debug \
    -DCMAKE_C_COMPILER=clang -DCMAKE_CXX_COMPILER=clang++
make -Cbuild -j12 tests
build/tests/tests  tests

```

## Types of changes

* Refactoring without change of behavior
* Preparing for new feature (non-breaking change which adds functionality)

## Checklist

- [x] Add tests to cover changes as needed.
- [x] I have read the [guidelines](https://developer.trustwallet.com/wallet-core/newblockchain#integration-criteria) for adding a new blockchain

PiRK commented:
I was not able to determine if `CashAddress` is part of the public interface of the library or not.

I have a preference for having the `CashAddress` be a prefix agnostic base class, and `BitcoinCashAddress` enforce the "bitcoincash" prefix. But if necessary, I can change this so that `CashAddress` remains the main BCH implementation.

Current option implementend in this PR:
```
                CashAddress 
                    |
|----------------------------------------|
|                                        |
|                                        |
BitcoinCashAddress                       ECashAddress
```

Alternative option that does not affect client code that might use `CashAddress` already:

```
            BaseCashAddress 
                    |
|----------------------------------------|
|                                        |
CashAddress                              ECashAddress
```

hewigovens commented on src/Bitcoin/CashAddress.cpp:
```diff
@@ -27,10 +24,10 @@ static const uint8_t p2shVersion = 0x08;
 static constexpr size_t maxDataSize = 104;
 
-bool CashAddress::isValid(const std::string& string) {
+bool CashAddress::isValid(const std::string& cashHRP_, const std::string& string) {
```
Just use `hrp`, no need to add `cash` prefix inside `CashAddress` class
```suggestion
bool CashAddress::isValid(const std::string& hrp, const std::string& string) {
```

catenocrypt replied:
Second to that; also please avoid using underscore prefix&suffix in parameter names.

PiRK replied:
OK. But I will need to find some way to disambiguate between the class member, the local variable (already called `hrp`) and the name of the parameter passed to the method.

catenocrypt replied:
As a rule of thumb, class members can always by disamiguated by `this->` if needed.
I prefer underscore if needed in local variable, not in param name.
But this is minor issue, not really relevant.

hewigovens commented on src/Bitcoin/CashAddress.h:
```diff
@@ -24,23 +26,26 @@ class CashAddress {
     std::array<byte, size> bytes;
 
+    /// Cash address human-readable part
+    const std::string cashHRP;
```
```suggestion
    const std::string hrp;
```

hewigovens reviewed (commented):
A generic hrp `CashAddress` looks good to me

catenocrypt commented on src/Bitcoin/CashAddress.h:
```diff
@@ -15,6 +15,8 @@
 
 namespace TW::Bitcoin {
 
+static const std::string BITCOIN_CASH_HRP = "bitcoincash";
```
If this is used by BitcoinCashAddress class, it's better to move inside that class.

PiRK commented:
I addressed the reviewer's  feedback: no trailing underscore in variable name, use `hrp` directly instead of `cashHRP` inside the class, 

I also renamed a local variable `hrp` to `decodedHRP` to avoid conflicting with the class member name. And I  decided to pass the `"bitcoincash"` prefix direclty as a literal string to the ctors instead of creating an unecessary new `BITCOIN_CASH_HRP` constant.

hewigovens commented on src/Bitcoin/CashAddress.h:
```diff
@@ -53,4 +56,23 @@ inline bool operator==(const CashAddress& lhs, const CashAddress& rhs) {
+  public:
+    /// Initializes an address with a string representation.
+    explicit BitcoinCashAddress(const std::string& string)
+        : CashAddress("bitcoincash", string) {}
```
can we add "bitcoincash" as a class static variable? please refer to `Zilliqa/Address.{h, cpp}` example

hewigovens reviewed (approved):
LGTM, only one comment left

PiRK commented:
I have a question about the next PR I'm working on, but related to this one. What is the script that I should use to add `HRP_ECASH` to TWHRP.h ?

I tried `codegen/bin/newcoin ecash` (after filling all necessary information including the hrp in `registry.json`, and it did generate a bunch of files and edited a few more, but it did not touch that particular TWHRP.h  file.

catenocrypt commented:
> I have a question about the next PR I'm working on, but related to this one. What is the script that I should use to add `HRP_ECASH` to TWHRP.h ?

`tools/generate-files` regenerates TWHRP.h
The enum is filled with values from the `"hrp"` fields in `registry.json`, see:
https://github.com/trustwallet/wallet-core/blob/master/registry.json#L448
